### ISSUE TITLE
[FEAT][Manifest] align elementwise multi-input ops (where / clamp / masked_fill) to PyTorch public API

### ISSUE THREAD

**Issue by lcy-seso** [tile-ai/TileOPs#1088]
### Symptom / Motivation
Discovered during PR #1081 ([Maintain][Manifest] add elementwise multi-input entries).

The four manifest entries added in #1081 narrow PyTorch's public API to the slice the current kernels implement:

- `ClampFwdOp` (`torch.clamp`) — scalar `min`/`max` only; Tensor `min`/`max` variant is BLOCKED at the spec level.
- `MaskedFillFwdOp` (`torch.Tensor.masked_fill`) — scalar `value` only; 0-dim Tensor `value` variant is BLOCKED.
- `WhereFwdOp` (`torch.where`) — `shape_rules` require equal shapes; PyTorch broadcasting is not modeled.

These narrowings are only documented inline in manifest comments / PR scope notes. Without a tracking issue they will be lost once #1081 merges and #1077 closes.

### Root Cause Analysis
Out of scope for #1081, which intentionally landed only the slice matching the existing kernel implementations under `status: spec-only`. The manifest must be aligned to PyTorch's public API regardless of current kernel coverage; kernel work is tracked separately.

### Related Files
- `tileops/manifest/elementwise.yaml` (entries `WhereFwdOp`, `ClampFwdOp`, `MaskedFillFwdOp`)
- `scripts/validate_manifest.py` (`_SHAPE_RULE_BUILTINS` helper allowlist)
- `tests/test_validate_manifest.py`
- `tileops/perf/formulas.py` (`where_fwd_roofline` and any new variant roofline funcs)

## Goal
Align the three multi-input elementwise op manifest entries to PyTorch's public API. Express each PyTorch-supported call form as an explicit manifest entry (split into `variant_of` entries where the manifest's "No `Optional[Tensor]`" rule requires it). All new / restructured entries land as `status: spec-only`; kernel and op-class conformance is deferred to follow-up PRs per the manifest trust model.

## Plan
**Plan type: confirmed**

### Naming convention (applies to Clamp and MaskedFill below)

Establish for variant families: **no suffix = all Tensor inputs (general form, primary); `Scalar` suffix = the parameter is `Number`/`None`**. The general Tensor form is the primary entry; scalar / partial-Tensor forms are `variant_of` it. Conceptually scalar is the degenerate case of Tensor (0-dim or constant broadcast), so the general form being primary matches "variant = specialization".

### 1. `scripts/validate_manifest.py` — add helpers

Extend `_SHAPE_RULE_BUILTINS` with two helpers used by the new shape_rules below. Equivalent expressions are not used; helpers are preferred for semantic clarity.

- `broadcast_shapes(*shapes) -> tuple` — bidirectional broadcasting, semantics aligned with `torch.broadcast_shapes`. Pure-Python implementation; do not import torch into the validator.
- `is_broadcastable_to(src, dst) -> bool` — unidirectional broadcastability check (can `src` broadcast up to `dst` without expanding `dst`).

Add unit tests in `tests/test_validate_manifest.py`: helper behavior + end-to-end via shape_rules eval.

#### 2.1 `WhereFwdOp` — broadcasting

- Replace the equal-shape `shape_rules` with:
  ```yaml
  - "out.shape == broadcast_shapes(condition.shape, input.shape, other.shape)"
  ```
- Keep `other` dtype as `same_as(input)`. Add a TODO comment: future dtype promotion is handled by an input cast in the kernel entry; the manifest does not introduce promotion semantics.
- Remove the inline BLOCKED note.

#### 2.2 Clamp family — split into 4 entries

PyTorch dispatcher exposes `aten::clamp` (scalar bounds) and `aten::clamp.Tensor` (Tensor bounds), plus public APIs `torch.clamp_min` / `torch.clamp_max`. Mapped to four manifest entries:

| Entry | Role | `ref_api` | Bounds |
|---|---|---|---|
| `ClampFwdOp` | primary | `torch.clamp` | both `min` and `max` required Tensor inputs |
| `ClampScalarFwdOp` | `variant_of: ClampFwdOp` | `torch.clamp` | `min`, `max` as `Number \| None` params |
| `ClampMinFwdOp` | `variant_of: ClampFwdOp` | `torch.clamp_min` | Tensor `min` only |
| `ClampMaxFwdOp` | `variant_of: ClampFwdOp` | `torch.clamp_max` | Tensor `max` only |

- `ClampScalarFwdOp` `min`/`max` change from `float | None` to `Number | None` (PyTorch accepts int).
- Tensor-bound entries use `broadcast_shapes(...)` for their `shape_rules`.
- All four `status: spec-only`. Existing `tileops/ops/elementwise.py:ClampFwdOp` (currently scalar form) will be aligned to `ClampScalarFwdOp` in a follow-up op PR; the new `ClampFwdOp` (Tensor form) is a new op to be implemented later.
- Add TODO: dtype promotion handled by kernel-entry cast in future, not in manifest.

#### 2.3 MaskedFill family — split into 2 entries + mask broadcasting

PyTorch dispatcher exposes `aten::masked_fill.Scalar` and `aten::masked_fill.Tensor` (0-dim Tensor `value`). Out-of-place `Tensor.masked_fill` returns a new tensor whose shape is the **bidirectional broadcast** of `input` and `mask` — verified empirically: `torch.zeros((2,1)).masked_fill(torch.zeros((2,3), dtype=torch.bool), 1.0)` returns shape `(2,3)`. (In-place `masked_fill_` is unidirectional, but the manifest models out-of-place only.)

| Entry | Role | `ref_api` | `value` |
|---|---|---|---|
| `MaskedFillFwdOp` | primary | `torch.Tensor.masked_fill` | 0-dim Tensor input (`shape: []`) |
| `MaskedFillScalarFwdOp` | `variant_of: MaskedFillFwdOp` | `torch.Tensor.masked_fill` | `Number` param |

- Both entries' `shape_rules`:
  ```yaml
  - "out.shape == broadcast_shapes(input.shape, mask.shape)"
  ```
  Bidirectional: `input` and `mask` broadcast against each other; output takes the broadcast shape.
- `MaskedFillScalarFwdOp` `value` type changes from `float` to `Number` (PyTorch accepts int/float/bool).
- All `status: spec-only`. Existing `tileops/ops/elementwise.py:MaskedFillFwdOp` (currently scalar form) will be aligned to `MaskedFillScalarFwdOp` in a follow-up op PR; the new `MaskedFillFwdOp` (Tensor `value`) is a new op to be implemented later.
- Add TODO: dtype promotion handled by kernel-entry cast in future.

### 3. Roofline updates (`tileops/perf/formulas.py` + manifest `roofline` blocks)

- `WhereFwdOp`: keep `func: tileops.perf.formulas.where_fwd_roofline`. Update the func to compute `N_total = product(out.shape)` (post-broadcast) instead of `product(input.shape)`.
- New Clamp variants: each declares its own `roofline` block. Sketch:
  - `vars.N_total: "product(out.shape)"`
  - `flops`: 1× compare-select for single-bound entries; 2× for both-bound entries.
  - `bytes`: read each present input + write output (Tensor bounds contribute `N_total * elem_bytes` per Tensor input under broadcasting; this assumes logical bytes — stays consistent with existing convention).
- `MaskedFillFwdOp` / `MaskedFillScalarFwdOp`: Tensor-`value` variant adds 1 byte for the 0-dim Tensor read (negligible vs N_total). `mask` byte cost stays at `N_total` (1 byte per logical output element after broadcast).

### 4. Validation

Run before PR open:

```
scripts/validate_manifest.py --levels schema
pytest tests/test_validate_manifest.py
```

### 5. Trust-boundary note

This PR touches both `tileops/manifest/` (manifest owner) and `scripts/validate_manifest.py` (validator owner). The two helper additions are introduced specifically to express the new shape_rules; landing them in separate PRs would leave dead helpers in the validator before manifest entries activate them. The boundary crossing is acknowledged and accepted in this PR as a one-shot exception, not a precedent.

## Constraints
- Must not regress PR #1081 (existing entries remain valid under L0 schema validation, modulo intentional restructuring documented above).
- Manifest changes go in their own PR per `.claude/rules/manifest-trust-model.md`. No `tileops/ops/`, `tileops/kernels/`, `tests/ops/`, or `benchmarks/` changes in this PR.
- No `Optional[Tensor]` in any new entry; all conditional Tensor inputs are split via `variant_of`.
- All new / restructured entries are `status: spec-only`.

## Acceptance Criteria
- [ ] AC-1: `WhereFwdOp.shape_rules` uses `broadcast_shapes(...)` and matches `torch.where` broadcasting semantics; inline BLOCKED note removed.
- [ ] AC-2: Clamp family expressed as 4 entries (`ClampFwdOp` primary + `ClampScalarFwdOp` / `ClampMinFwdOp` / `ClampMaxFwdOp` variants) with `Number | None` for scalar params; inline BLOCKED note removed.
- [ ] AC-3: MaskedFill family expressed as 2 entries (`MaskedFillFwdOp` primary + `MaskedFillScalarFwdOp` variant) with bidirectional `input`/`mask` broadcasting (out-of-place `Tensor.masked_fill` semantics) and `Number` value type; inline BLOCKED note removed.
- [ ] AC-4: `_SHAPE_RULE_BUILTINS` includes `broadcast_shapes` and `is_broadcastable_to`, with unit tests in `tests/test_validate_manifest.py`.
- [ ] AC-5: `scripts/validate_manifest.py --levels schema` passes for all affected entries.
- [ ] AC-6: `where_fwd_roofline` and new variant roofline blocks are consistent with the post-broadcast `N_total` convention.
- [ ] AC-7: Follow-up issue filed tracking kernel + op-class conformance to all new / restructured entries.
